### ISSUE TITLE
Taking gradients through jnp.arctan2 seems to be very unstable

### ISSUE THREAD

**Issue by Peter-Vincent** [google/jax#15407]
### Description

I'm trying to optimise a model and I'm finding that the gradients are very very unstable - no learning rate lets me reliably move in the right direction to reduce the loss.  I have narrowed down the instability to the line where I call `jnp.arctan2`  to calculate angles.

I have included below a minimum working example  which demonstrates the issue.  This may well be a misunderstanding on my behalf of the behaviour of `jnp.arctan2`.

```
import jax.numpy as jnp
from jax import value_and_grad, random

def f(x,mu):
    p_x  = jnp.exp(5 * (jnp.cos(jnp.arange(x.shape[1]) - mu) - 1))
    real = jnp.cos(x) @ p_x
    imag = jnp.sin(x) @ p_x
    return jnp.sum(jnp.arctan2(imag,real)) # Change the commented return value to explore behaviout w/ and  w/o arctan2
    #return jnp.sum(imag + real)

def opt_f(num_iter = 10,lr=1e-3):
    key = random.PRNGKey(0)
    x   = (random.uniform(key,shape=(100,100)) * 2 * jnp.pi) - jnp.pi
    mu  = 0.0
    grad_f = value_and_grad(f,1)
    i = 0
    while i < num_iter:
        f_val,f_grad_x = grad_f(x,mu)
        mu = mu - lr * f_grad_x
        print(f"iteration {i} value {f_val}")
        i += 1

if __name__ == "__main__":
    opt_f()
```

If I run this code as is, then the first 10 iterations descend as expected (see below)
```
iteration 0 value -40.33513641357422
iteration 1 value -40.38768768310547
iteration 2 value -40.43447494506836
iteration 3 value -40.47615051269531
iteration 4 value -40.51304626464844
iteration 5 value -40.545711517333984
iteration 6 value -40.5745849609375
iteration 7 value -40.60002899169922
iteration 8 value -40.62242889404297
iteration 9 value -40.64210510253906
```
But if I change the return value to include the conversion to angle, then the gradients are very unstable and the output fluctuates wildly
```
iteration 0 value 16.434661865234375
iteration 1 value 16.08127784729004    # Down
iteration 2 value 15.644673347473145
iteration 3 value 21.444284439086914
iteration 4 value 20.929534912109375
iteration 5 value 26.625213623046875
iteration 6 value 19.6180419921875
iteration 7 value 12.849037170410156
iteration 8 value 19.092792510986328
iteration 9 value 19.084476470947266
```
Reducing the learning rate of course minimises the size of the fluctuations, but the directions still vary from iteration to iteration, see below
```
iteration 0 value 16.434661865234375
iteration 1 value 16.434650421142578
iteration 2 value 16.434680938720703
iteration 3 value 16.434490203857422
iteration 4 value 16.43451690673828
iteration 5 value 16.434511184692383
iteration 6 value 16.434545516967773
iteration 7 value 16.43435287475586
iteration 8 value 16.43438720703125
iteration 9 value 16.434375762939453
```

### What jax/jaxlib version are you using?

0.4.6

### Which accelerator(s) are you using?

CPU

### Additional system info

Python 3.10.8, Ubuntu 20.04.5 LTS in WSL2
